Add support for listing struct/union/enum definitions using BTF

mmisono opened the pull request:
'-l' option now can take struct/union/enum name and dump its definition
using BTF. For example:

```
% sudo ./src/bpftrace -l "enum bpf*"
enum bpf_access_type
enum bpf_adj_room_mode
enum bpf_arg_type
enum bpf_attach_type
enum bpf_cgroup_storage_type
[...]
```

and

```
% sudo ./src/bpftrace -lv "struct path"
BTF: using data from /sys/kernel/btf/vmlinux
struct path {
        struct vfsmount *mnt;
        struct dentry *dentry;
};
```

##### Checklist

- [x] Language changes are updated in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`

mmisono commented:
@olsajiri 
I think the second commit https://github.com/iovisor/bpftrace/pull/1340/commits/c5959b6a2451474ff633fe7a035bf727a9a8a7e3 is harmless, but could you take a look when you can?

danobi commented on src/list.cpp:
```diff
@@ -286,9 +258,79 @@ void list_probes(const BPFtrace &bpftrace, const std::string &search_input)
+  }
+
+  // software
+  if (has_wildcard_in_probe || probe_name.empty() || probe_name == "software")
```
Maybe pull out `has_wildcard_in_probe || probe_name.empty()` into another variable to repeat less code

danobi commented on src/btf.cpp:
```diff
@@ -584,6 +584,87 @@ void BTF::display_funcs(std::regex *re) const
+    std::string def = c_def(struct_set);
+    // c_def() contains all the necessary dependent types needed for
+    // compilation. Print definition of given structs (or union/enum) only
+    for (auto &name : vec)
```
```suggestion
    for (const auto &name : vec)
```

danobi commented on src/btf.cpp:
```diff
@@ -584,6 +584,87 @@ void BTF::display_funcs(std::regex *re) const
+      auto end = std::min(def.find("\n};", start) + 3, def.size());
+      if (start == std::string::npos)
+        continue;
+      for (auto p = def.begin() + start; p < def.begin() + end; p++)
+      {
+        std::cout << *p;
+      }
```
Feel a bit strange to use iterators for this. Why not string index and chars?

ie
```
for (size_t i = start; start < end; ++i)
  std::cout << def[i];
```

mmisono commented:
@danobi 
Thanks for the review. I fixed the reviewed points. I also added https://github.com/iovisor/bpftrace/pull/1340/commits/9a4757935ecb5291acd92667426c90cb543d1d5d to list only usdt when a pid is given. For example:

```
% sudo ./src/bpftrace -lv -p 25655
BTF: using data from /sys/kernel/btf/vmlinux
usdt:/proc/25655/root/lib/x86_64-linux-gnu/libc-2.27.so:libc:setjmp
usdt:/proc/25655/root/lib/x86_64-linux-gnu/libc-2.27.so:libc:longjmp
usdt:/proc/25655/root/lib/x86_64-linux-gnu/libc-2.27.so:libc:longjmp_target
[...]
```

Please let me know if this has a problem.

danobi commented on src/list.cpp:
```diff
@@ -286,9 +258,81 @@ void list_probes(const BPFtrace &bpftrace, const std::string &search_input)
   // kfuncs
-  bpftrace.btf_.display_funcs(search.empty() ? NULL : &re);
+  if (list_all || probe_name == "kfunc")
+    bpftrace.btf_.display_funcs(search.empty() ? NULL : &re);
```
style nit
```suggestion
    bpftrace.btf_.display_funcs(search.empty() ? nullptr : &re);
```

danobi commented:
> Please let me know if this has a problem.

Seems like a reasonable change

mmisono commented:
- Updated changelog and fixed nits.
